Opencv agast tree search proposal for significant codesize reduction with comparative performance

askourik opened the pull request:
Hi All,
I propose to fix the https://github.com/Itseez/opencv/issues/5669 problem and reduce codesize for AGAST algorithm up to 5-10 times depending on platform. It can be very useful for embedded systems with small memory resources because current AGAST codesize is one of largest in OpenCV. The performance remains comparative (50-200% variation depended on testcase and method). The core of improvements is binary tree condition search from 32-bit table instead of thousands of if-else-goto in previous code. The 32-bit tree leaf format contains two 12-bit absolute table offsets for positive/negative results, 4-bit id of compared node/pixel (can be expanded up to 7 bit if necessary) and 1 bit of comparing result (below or above threshold). The leaf table was generated automatically from current code using Perl script, all performance/ quality tests passed OK. The current proposed branch contains code replacement, but it's easy to derive new class if necessary. Also I have experimental version with even more table codesize reduction (24-bit instead of 32-bit plus relative offsets), but it seems to be slower on RISC processors because of alignments. The further performance optimization is also possible after tree reordering for better cache alignmants (I precerved the current order). Feel free to ask me any questions and proposals about this improvement.
Regards,
Andrey

cbalint13 commented:
@askourik , 

* I like your proposal, very nice optimisation !
* Can please add the perl code too somewhere in some sample folder ?

For this to get merged, better would be to (at least i would do):
(a) leave the original code too, new code should be conditioned with some flag.
(b) need to check in the depending BRISK if performance is degraded or not.

I help you with (b) and will look how we should do (a) properly.

--------------<<<--------------

Hmm i do not understand why we get there those warnings, i guess buildbot's fault and will get fixed soon, Cc @mshabunin ,

askourik commented:
@cbalint13 ,
Of course, I can add the appropriate Perl code, but not familiar with OpenCV policy on this topic. Could you please tell me correct path to put  my script? To be mentioned: it requires some manual correction of existing code, you should split 1-line "else if" construction in 2 lines. Also, you shouls set first and last lines of each "if-else-goto" large module as input parameter. I can provide brief instructions and batch file with working examples if necessary.

On other suggestions:
a) yes, I have merged version with AGAST_TREE preprocessor define which switches on/off this functionality, If it's the best way, I can do atother pull request, or you mean software-defined flag?
b) I guess the performance is deeply depends on test platform. I tested for Intel, In general it 1.5-2 times slower for current version. But due to regular structure, it can be easily improved by tree reordering an128-bit optimizers as well as using close memory pools for pixel/tree data
BTW, what about 24-bit tree version?
Regards,
Andrey

cbalint13 commented:
@askourik ,

- I propose you something like: http://openrisc.rdsor.ro/AGAST/agast_trees.diff I have no better idea but to leave the original for x86_64/i386 ones. The very intention of the original authors for AGAST was those decision trees, they stated: "faster than FAST and even FAST-ER" http://www6.in.tum.de/Main/ResearchAgast :) Thats why i have in mind to leave at least for x86_64/i386 their orignal code and idea.

- It's 1:1 your work i just added apropiate if/else/endif for x86 CPU, nothing else (runs&compiles).

- The quality of AGAST scoring routines is very good compared to FAST e.g BRISK detector/descriptor depends a lot on AGAST's scoring.

- I believe we can create a doc folder inside /feature2d (just like other module have such), lets put there agast_tree_gen.pl.txt (lets have .txt ending to not confuse with some exec) and in the very features2d/include/opencv2/features2d.hpp AGAST doxygen sections lets mention that for non i386/x86_64 there are some optimised tree based routines (bit slower) and a perl generator exists for it (if someone want further work on it), and we can mention this PR&issue ticket and scope too there.

* Regarding 24bit i dont know :) You done an amazing optimization already ! Probably a furthermore even better approach would be to convert your tree stuff to use inline ASM for x86 and separate inline ASM for AMD like many algorithms does this (e.g FAST). Probably even original "goto" if one rewrites it in ASM would work even better. Perhaps someone could pick up your perl generator and tackle with ASM, thats why i proposed to push perl script too for a better transparency.


Lets see upstream reviewers opinion too, hope buildboot is fixed too in the meanwhile. I believe for non-intel you already done the great :)

askourik commented:
@cbalint13 ,
It's a good idea to leave original code for x86, I'll update my repositary tomorrow.
Yes, the brief documentation is necessary for new method because tree node structure needs description. Of course, the tree node internal format can be changed for (possible) further expansion on larger pixel areas, it's a discussion topic. Using perl script, it's a simple task.
OK, I'll create doc folder inside features2d and put here a family of perl scripts with batch file example for automation. But for good case reproduction, I'm also going to put there modified agast.cpp and agast_score.cpp files (renamed by *.txt). For original function OAST9_16, e.g., the "if-else-goto" structure is changed by "{;}-continue", so I manually replaced these operators by goto labels to simplify perl automation.
If anybody will need more compact 24-bit tree version, I can also provide appropriate Perl Script. It works only for agast.cpp because the distance between nodes in agast_score.cpp is too large (>256)
Feel free to add any comments before my tomorrow commit :)
Regards,
Andrey

askourik commented:
I updated AGAST code and added doc folder with perl scripts as promised.

cbalint13 commented:
@askourik , Very nice work !

* Last requirements, mandatory & cosmetic related:

(1) Can please remove any tabs and replace with spaces in the code ?

(2) Put final newline on that .bat script (its missing): https://github.com/Itseez/opencv/pull/5670/files#diff-93c8deac88ec5648d5a1a88d4284bd75R32

(3) In features2d/include/opencv2/features2d.hpp, mention that on non-intel there is tree optimised variant of AGAST with same numerical results, and also mention the presence of perl script.

(4) Squash all commits into single one so mainteiners/reviewers can read well your code.

-------------

My +1 comments:

* However those "register" https://github.com/Itseez/opencv/pull/5670/files#diff-28fe7c6104c1d3e31066fa5339283e80R8156 don't know how would perform on non-intel arch that maybe have limited set of registers, so your optimisation is really welcome for any non-intel. The BRISK which depends on AGAST seems to be not affected, at least not non-intel.

* Perl script looks very intuitive, anyone can do ASM optimize if want further improvements.

* @askourik , I cant see the buildbot build for now (you only pushed the perl code to doc so no rebuild was triggered since there was no toch to the very code). If do the mandatory cosmetics you will anyway trigger buildbot too. Its trivial to have this builded by buildbot and passed with buildbot's QA checks too (full green states).

askourik commented:
@cbalint13 ,
I've replaced tabs by spaces, put .bat newline and added description to features2d.hpp. Alas, the squashing all commits into one using Git is impossible task for me. I've tried to do it honestly, but no success, now the log is even more complicated :(. So I guess the best way in this situation is to clone final code on new branch and create new pull request for main repository. I'd really appreciate for any ideas in this situation.
The new buildbot build was passed as previous, iOS build still failing...

cbalint13 commented:
@askourik , try this way to reset to HEAD and force push (i hope i am correct):

1) Make safe-copy: ```wget https://github.com/Itseez/opencv/pull/5670.patch```
2) Make a fresh ```git clone https://github.com/askourik/opencv.git```
3) Go inside ```cd opencv```
4) Update with upstream ```git pull https://github.com/Itseez/opencv.git```
5) Push updates (if any updates at 4) ```git push```
5) Switch to agast_tree ```git checkout opencv_agast_tree```
6) Reset HEAD to master ```git reset --hard origin/master```
7) Now re-apply patch ```patch -p1 < somepath/5670.patch```
8) Bow commit like you did first time ```git commit -a -m "Add treelookups for nonintel"```
9) Force push whole ```git push --force```

* iOS failing due to fact buildbot is still not fully functional, i think they are in some upgrade/extend process. Dont worry.

askourik commented:
@cbalint13 ,
Thanks, the second attempt was more successful, but the doc folder wasn't added after this patch apply. I've added docs folder be separate commit, hope that now all is OK
Regards,
Andrey

vpisarev commented:
@askourik, thank you for the patch!

could you please fix the warnings reported by the Android and iOS build slaves?
http://pullrequest.opencv.org/#/summary/

also, this is quite radical refactoring of the algorithm, so some regression tests would be useful to have too.

I do not have any major complains about the code; one of possible improvements would be to change "unsigned long" that you use extensively with "unsigned". "unsigned long" is 32-bit or 64-bit number, depending on the platform (e.g. it's 64-bit number on quite popular nowadays x64 Linux distros and also on OSX), whereas "unsigned" is 32-bit number of any platform and any C/C++ compiler I can remember. So those tables would be 2x more compact if you switch from "unsigned long" to "unsigned".

askourik commented:
@vpisarev ,
OK, I'll fix the Android/iOS bugs, it just require the declaration of one new function, I've lost it after trying of squashing commits all-in-1.
Regretfully, I cannot agree with your remarks on type of tree tables. According to C99, the "unsigned" type (brief form of "unsigned int") cannot guarantee 32-bit representation because it capable of containing at least the [−32767, +32767] range; thus, it is at least 16 bits in size (ISO/IEC 9899:1999 specification, TC3 (PDF). p. 22); moreover, all C data types can guarantee only minimal bit number because of portability; the "unsigned long" type guarantees at least 32 bits; all other problems are up to platform-dependent compiler. There is a special type "unsigned long long" which guarantees at least 64 bit; so, the correct x64 compiler must distinct the above two types and optimize them in different ways.
Regarding additional regression tests: I can only support this idea, but I need more details. Are you suggest to add new testcases in current test application or run it locally? Some new tests will be useful. But main "code-to-tables" conversion was automated using Perl script by the way similar to compiler code parsing; warnings and errors in Perl scripts are also provided. So the "human factor" error risk was minimized.
Which image samples from standard opencv test set could you recommend for additional testing?
Regards,
Andrey

alalek commented:
> According to C99, the "unsigned" type (brief form of "unsigned int") cannot guarantee 32-bit representation
In this case OpenCV doesn't not work at all. There are many things silently assume 32-bit "int".

What is about uint32_t?

askourik commented:
@alalek ,
I guess your suggest is reasonable, I'll implement it after fixing iOS/Android compile problem

askourik commented:
The iOS/Android bug fix is ready now. The unsigned long/uint32_t replacement was also made and works OK; the only open question is new regression tests addition, any recommendations on it?
Also, my first try to squash all-in-1 commits in opencv_agast_tree branch was very unsuccessful. If you still need it, I can generate new branch and create new pull request.
Regards,
Andrey

vpisarev commented:
@askourik, @alalek, thanks for solving the datatype issue!
@askourik, regarding the tests. As I see, there is already some test for AGAST: opencv/modules/features2d/test_agast.cpp. If it passes, then may be it's enough, no extra work needed

askourik commented:
@vpisarev , yes, there are bost performance and quality tests for AGAST algorithm, and I've checked both of them before pull request (see above discussion with @cbalint13).
Also, I can propose for OpenCV very fast and efficient Hough Transform algorithm which overtops all other Hough algorithms by performace/resolution ratio (and uses MMX/SSE optimizations). If you are interested and need details, let me know where to initiate new discussion about it.
Regards,
Andrey

cbalint13 commented:
@askourik 

  * on very ARM can check if running opencv/modules/features2d/test_agast.cpp & features2d/perf/perf_agast.cpp passes ? GitHub's buildbot does not do these checks on very ARM.
  * on Intel it passes since we use ```#ifdef``` and old code is running.

askourik commented:
@cbalint13 ,
Yes, it passes on my Raspberry Pi 2. I can re-run these tests using files from my last commit and provide you brief results in this discussion or elsewhere.
P.S. Are there any possibilities to check performance/regression tests using Github buildbot for ARM platform?

cbalint13 commented:
@askourik , 

Buildbot looks green now, so it could be merged.

* As @vpisarev also stated if it **pass** we are on the fine path.
* We should ask @mshabunin if ARM could perform tests too on buildbot ? 

I guess buildbot's ARM is not that perfromant to afford running the tests too. Thank you for spending time on this PR and doing ARM tests.

----<<<----
* For the MMX/Hough thing would be great to open a PR .Notice that there already is a hogh transform implemented, but i believ it can be enhanched. I plan soon to ad Radon transform among the hough ones so your MMX optimization comes just fine.

askourik commented:
@cbalint13, @vpisarev,
Here are brief test logs from Raspberry Pi 2 (new AGAST module):
./opencv_test_features2d
CTEST_FULL_OUTPUT
OpenCV version: 3.0.0-dev
OpenCV VCS version: 3.0.0-783-g7ddc2cc-dirty
Build type: release
Parallel framework: openmp
CPU features: neon
OpenCL is disabled
[==========] Running 86 tests from 45 test cases.
[----------] Global test environment set-up.
[----------] 1 test from Features2d_AGAST
[ RUN      ] Features2d_AGAST.regression
[       OK ] Features2d_AGAST.regression (1623 ms)
[----------] 1 test from Features2d_AGAST (1623 ms total)
...
[----------] 1 test from Features2d_Detector_AGAST
[ RUN      ] Features2d_Detector_AGAST.regression
[       OK ] Features2d_Detector_AGAST.regression (685 ms)
[----------] 1 test from Features2d_Detector_AGAST (685 ms total)
...
./opencv_perf_features2d
...
[==========] Running 59 tests from 16 test cases.
[----------] Global test environment set-up.
[----------] 8 tests from agast_detect
[ RUN      ] agast_detect.detect/0
[ PERFSTAT ]    (samples = 10, mean = 78.81, median = 78.79, stddev = 0.10 (0.1%))
[ VALUE    ] 	("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", AGAST_5_8)
[       OK ] agast_detect.detect/0 (954 ms)
[ RUN      ] agast_detect.detect/1
[ PERFSTAT ]    (samples = 10, mean = 81.89, median = 81.85, stddev = 0.11 (0.1%))
[ VALUE    ] 	("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", AGAST_7_12d)
[       OK ] agast_detect.detect/1 (984 ms)
[ RUN      ] agast_detect.detect/2
[ PERFSTAT ]    (samples = 10, mean = 82.21, median = 82.22, stddev = 0.10 (0.1%))
[ VALUE    ] 	("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", AGAST_7_12s)
[       OK ] agast_detect.detect/2 (987 ms)
[ RUN      ] agast_detect.detect/3
[ PERFSTAT ]    (samples = 10, mean = 84.31, median = 84.30, stddev = 0.09 (0.1%))
[ VALUE    ] 	("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", OAST_9_16)
[       OK ] agast_detect.detect/3 (1008 ms)
[ RUN      ] agast_detect.detect/4
[ PERFSTAT ]    (samples = 10, mean = 66.69, median = 66.66, stddev = 0.08 (0.1%))
[ VALUE    ] 	("stitching/a3.png", AGAST_5_8)
[       OK ] agast_detect.detect/4 (811 ms)
[ RUN      ] agast_detect.detect/5
[ PERFSTAT ]    (samples = 10, mean = 73.92, median = 73.87, stddev = 0.25 (0.3%))
[ VALUE    ] 	("stitching/a3.png", AGAST_7_12d)
[       OK ] agast_detect.detect/5 (883 ms)
[ RUN      ] agast_detect.detect/6
[ PERFSTAT ]    (samples = 10, mean = 74.48, median = 74.43, stddev = 0.21 (0.3%))
[ VALUE    ] 	("stitching/a3.png", AGAST_7_12s)
[       OK ] agast_detect.detect/6 (888 ms)
[ RUN      ] agast_detect.detect/7
[ PERFSTAT ]    (samples = 10, mean = 78.64, median = 78.60, stddev = 0.26 (0.3%))
[ VALUE    ] 	("stitching/a3.png", OAST_9_16)
[       OK ] agast_detect.detect/7 (929 ms)
[----------] 8 tests from agast_detect (7444 ms total)

askourik commented:
@vpisarev , @cbalint13 ,
Are there remains any problems or difficulties to merge my pull request? If yes, please let me know how to repair it. I guess that 3+ times code reduction for one of the largest openCV modules is a good reason to increase the merge priority of my pull request.
Regards,
Andrey

vpisarev commented:
sorry for delay! :+1:
